golangServer/api/import_by_url.go

```diff
@@ -0,0 +1,144 @@
+package api
+
+import (
+	"Oracle.com/golangServer/Oracle"
+	"Oracle.com/golangServer/config"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+)
+
+func ImportEventListener() {
+	// Import channels for logs
+	Logs := make(chan *Oracle.OracleImportFromUrl)
+	// Subscribe to each event
+	opts := &bind.WatchOpts{Context: context.Background(), Start: nil}
+	eventSub, err := config.OracleContract.WatchImportFromUrl(opts, Logs)
+	if err != nil {
+		log.Fatal("Failed to subscribe to Import events:", err)
+	}
+	// start Listening...
+	log.Println("ImportEvent Listening ...")
+	for {
+		select {
+		case err := <-eventSub.Err():
+			log.Println("[Error in Event IMPORT:", err)
+		case event := <-Logs:
+			log.Println("Received put event ", event.ReqID)
+			importByUrl(event)
+		}
+	}
+}
+
+// Import data to memory db
+func importByUrl(event *Oracle.OracleImportFromUrl) {
+	var statement bool
+	tps := GenTransactOpts(config.GasLimit)
+
+	colName := event.ColName
+	ctx := context.Background()
+	dbName := event.DbName
+	db := config.Dbs[dbName]
+	dbC, err := db.Collection(colName, "")
+	if err != nil {
+		log.Println("Get collection ERROR: ", err)
+		info := fmt.Sprintf("Get collection ERROR: %v", err)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+
+	resp, err := http.Get(event.Url)
+	defer resp.Body.Close()
+	if err != nil {
+		log.Println("Get datas ERROR: ", err)
+		info := fmt.Sprintf("Get datas ERROR: %v", err)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+	if resp.StatusCode != 200 {
+		log.Println("Get datas Fail, StatusCode = ", resp.StatusCode)
+		info := fmt.Sprintf("Get datas Fail, StatusCode = %v", resp.StatusCode)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+	if resp.Header.Get("Content-Length") == "0" {
+		info := fmt.Sprintf("Empty content, Content-Length = 0")
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+
+	switch event.Format {
+	case "csv":
+		body, err := io.ReadAll(resp.Body)
+		reader := strings.NewReader(string(body))
+		err = IngestCSV(ctx, reader, dbC)
+		err = db.ApplyChanges(ctx)
+		if err != nil {
+			log.Println("Unmarshal csv ERROR: ", err)
+			info := fmt.Sprintf("Unmarshal csv ERROR: %v", err)
+			statement = false
+			//response to oracle
+			config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+			return
+		} else {
+			statement = true
+		}
+	case "ndjson":
+		body, _ := io.ReadAll(resp.Body)
+		jsonArrayStr := string(body)
+
+		cleanedStr := strings.Trim(jsonArrayStr, "[]")
```

Comment:
This is because the JSON data passed by the client at that time, now when the client passes data in NDJSON format, this code snippet is not needed, it has been corrected.